File: plot_scripts/mutation_types.R
source("plot_scripts/setup.R")

################################################################################
# resampling
muta_data <- as.data.frame(read.csv(file = "plot/tmp_data/mutation_types.csv"))

p <- muta_data %>%
    mutate(not_found = covered - found, not_covered = done - covered) %>%
    mutate(Found = found / done, Covered = not_found / done) %>%
    mutate(pattern_name = gsub("_", " ", pattern_name)) %>%
    filter(covered > 0) %>%
    # pivot_longer(cols = c(not_covered, not_found, found)) %>%
    pivot_longer(cols = c(Found, Covered)) %>%

    ggplot(aes(x = fuzzer, y = value, fill = name)) +
    facet_wrap(~pattern_name, labeller = label_wrap_gen(width = 15)) +
    scale_y_continuous(labels = scales::percent) +
    geom_col() +
    theme(
        legend.position = "bottom",
        axis.title.x = element_blank(),
        strip.text = element_text(size = 6),
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)
    ) +
    labs(y = "Percentage of Mutants", fill = "Result")

ggsave(p, filename = "plot/fig/mutation_types.pdf", device = "pdf", width = 5, height = 7)

File: plot_scripts/mutation_types.py
#%%
import sqlite3
from helper import db_connect, fix_path, query, to_latex_table, out_path, data_path

db = db_connect("data/current/stats_all.db")


#%%
sampling_results = query(db, """
SELECT pattern_name, fuzzer, sum(done) AS done, sum(covered) AS covered, sum(found) AS found
FROM run_results_by_mut_type_and_fuzzer
JOIN mutation_types USING (mut_type)
GROUP BY mut_type, fuzzer;
""")

#%%
all_runs_list = []

for run in sampling_results:
    pattern_name = run.get("pattern_name")
    fuzzer = run.get("fuzzer")
    done = run.get("done")
    covered = run.get("covered")
    found = run.get("found");
    all_runs_list.append({
        'pattern_name': pattern_name,
        'fuzzer': fuzzer,
        'covered': covered,
        'found': found,
        'done': done,
    })

#%%
import pandas as pd
data = pd.DataFrame.from_records(all_runs_list)

data.to_csv("plot/tmp_data/mutation_types.csv")
